### PR TITLE
Default log entry label providers

### LINKED ISSUE #2120: Default log entry labels with HTTP context info
Would you consider a default `ILogEntryLabelProvider` implementation (or multiple) which populate some labels related to an HTTP request useful for searching and correlating logging entries? It could be registered when you use `UseGoogleDiagnostics()` by default. I was thinking of the following values:

- The HTTP [`TraceIdentifier`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.traceidentifier?view=aspnetcore-2.0)
- The HTTP user identity name 
   - Might be a privacy issue, perhaps make configurable or just whether a request was authenticated
- Grab the exception from [`IExceptionHandlerPathFeature`](https://docs.microsoft.com/nl-nl/dotnet/api/microsoft.aspnetcore.diagnostics.iexceptionhandlerpathfeature?view=aspnetcore-2.0) (when available) and add the exception type/message

I find myself adding this provider to all of my projects hosted in GCP:

```cs
public void Invoke(Dictionary<string, string> labels)
{
    var httpContext = _httpContextAccessor.HttpContext;
    if (httpContext == null)
    {
        // Not in the scope of an HTTP request
        return;
    }

    // Add the trace identifier to correlate log entries from a single request
    labels["trace_identifier"] = httpContext.TraceIdentifier;

    // Add authenticated user info
    if (httpContext.User?.Identity?.IsAuthenticated == true)
    {
        labels["user_authenticated"] = bool.TrueString;
        labels["user_name"] = httpContext.User.Identity.Name;
    }

    // Add any exception info when available to quickly inspect 
    // and filter exceptions in the GCP Logging console
    var exceptionFeature = httpContext.Features.Get<IExceptionHandlerPathFeature>();
    var exception = exceptionFeature?.Error;
    if (exception != null)
    {
        labels["error_path"] = exceptionFeature.Path ?? "N/A";
        labels["error_type"] = exception.GetType().FullName;
        labels["error_message"] = exception.Message;
        labels["error_stacktrace"] = exception.ToStringDemystified();
    }
}
```

What are your thoughts about this?

cc @iantalarico

### PR DESCRIPTION
**Pull request by henkmollema**
Adds default log entry label providers which a user can add. The `TraceIdLogEntryLabelProvider` configured by default. The others are opt-in.

Resolves #2120

### CODE REVIEW

**Comment by henkmollema** [comment]
@iantalarico this is a work-in-progress. I'm wondering what you think of the general approach.

Beside the providers listed in #2120, these ones might be valuable too:
- The client IP address
- The ASP.NET Core environment name

What do you think?

**Comment by iantalarico** [comment]
@henkmollema Sorry for the slow reply was in London and traveling.  I will take a look at this today.

**Comment by iantalarico** [comment]
@henkmollema I very much like these, especially the environment name (I think this will come in handy).

Ping me when you're ready for a review.  Again thank you so much for the continued PR, it's great to see the community involvement.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/LabelProviders/LabelProviderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        /// <summary>
+        /// Adds a <see cref="ILogEntryLabelProvider"/> of type <typeparamref name="T"/> to the service collection instance.
+        /// </summary>
+        /// <typeparam name="T">The type of the <see cref="ILogEntryLabelProvider"/> implementation.</typeparam>
+        /// <param name="serivces">The <see cref="IServiceCollection"/> instance.</param>
+        /// <returns>The <see cref="IServiceCollection"/> instance.</returns>
+        public static IServiceCollection AddLogEntryLabelProvider<T>(this IServiceCollection serivces)
+            where T : class, ILogEntryLabelProvider
+        {
+            return serivces.AddSingleton<ILogEntryLabelProvider, T>();
```
[the remark this answers is not in the document]
Comment (henkmollema) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/LabelProviders/LabelProviderExtensions.cs]: Of course, I like that too.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/LabelProviders/UserLogEntryLabelProvider.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+using System.Security.Claims;
+using System.Security.Principal;
+using Microsoft.AspNetCore.Http;
+
+namespace Google.Cloud.Diagnostics.AspNetCore
+{
+    /// <summary>
+    /// A <see cref="ILogEntryLabelProvider"/> which adds the information of the authenitacted user to the log entry labels.
+    /// </summary>
+    public class UserLogEntryLabelProvider : HttpLogEntryLabelProvider
```
[the remark this answers is not in the document]
Comment (iantalarico) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/LabelProviders/UserLogEntryLabelProvider.cs]: Well in that case I guess it's okay :)

**Comment by henkmollema** [comment]
Awesome! I think I'm all set for merging this when the build is green. I've not run the integration tests locally so there's a chance they fail. 😆

**Comment by iantalarico** [comment]
@henkmollema I'll give them a run before I merge.  Feel free to ask me to do this in the future as I know it can be a pain and depending on the tests can have a cost associated with it.

**Comment by henkmollema** [comment]
> @henkmollema I'll give them a run before I merge. Feel free to ask me to do this in the future as I know it can be a pain and depending on the tests can have a cost associated with it.

Good to know, thanks. 've almost spent my free credit. 😄 

> Just one small change I missed (that the integration tests caught)

Ah, fixed it.

**Comment by iantalarico** [comment]
@henkmollema Yea I'm not sure if the logs from these tests would cost but they have the potential too.

Thanks for the fix!

**Comment by henkmollema** [comment]
Thanks!

**Comment by iantalarico** [comment]
@henkmollema This is now in release 3.0.0-beta05

**Pull request merged**